### PR TITLE
Parallelization of score and diversity calculation

### PR DESCRIPTION
**Pull request by NPException**
I tried to do the parallelization where it makes immediate sense. I make use of the fancy Java 8 features for that, so I hope you build javolver with a Java 8 JDK :)

If you have any questions on how the stuff works, do not hesitate to ask.

### CODE REVIEW

**Comment by NPException** [comment]
The .gitignore change just contains files that IntelliJ IDEA placed into the project on my PC.

File: javolver/JavolverRanking.java [hunk 1]
```diff
@@ -34,7 +37,7 @@ public int compare(Individual ind1, Individual ind2)
 	
 	public static void calculateDiversityRank(ArrayList<Individual> pool) {
 		// Calculate the average chromosome from the pool.
-		Chromosome averageChromosome = CalculateAverageChromosome(pool);
+		Chromosome averageChromosome = calculateAverageChromosome(pool);
```
Comment (NPException) [inline javolver/JavolverRanking.java]: Sorry for that one, I had to do it :D

**Comment by nickd3000** [comment]
Hi!, Thanks for submitting that :)

I've noticed a couple of problems though, sorry!

I think the bulk of the processing load in the system would actually be from scoreGenes() - this calls every individual in the pool and gets them to perform the fitness function which will tend to be the slowest thing (eg. In the tree example it will recursively build the tree and calculate the score of that tree geometry) - the ones you parallelised (unfortunately) are more of a sorting pass on the individuals that ranks them in order of the scores calculated from scoreGenes().  This ranking is done so that there is not such a huge difference between the scores of good and bad individuals (and it is only used if the config is set to use ranking).  So I don't think the sorting part is going to give as much gain from parallelising as would the scoring mechanism.

So my suggestion, if it's possible, is to make a version of scoreGenes() that uses threads like you did for the ranking functions, and then we can keep the doOneCycle function a bit cleaner by just calling that new function depending on the config settings. (I would drop the changes you made to the ranking functions)

It would be good to put the new parallel scoring function in a new class too, maybe called JavolverParallel and the method called scoreGenesParallel?

One other thing - if this could be done without java 8 features that would be a bonus - it's fun for us to use but it can be a pain for some novices to get running (especially on linux), so I'd prefer not to rely on java 8 unless theres a really good beneficial reason :)


Many thanks for contributing and I hope this didn't sound too negative, but let me know what you think about the proposed changes.

Nick

**Comment by NPException** [comment]
So much text :D

Okay, yes that makes sense. Depending on what the user is actually doing in the `Individual.calculateScore()` method, the scoring will indeed be the most processing intense part.

Moving the parallel scoring into a separate class sounds a bit overkill though. I'd suggest to use a simple config switch in `doOneCylce()`:
```java
if (config.parallelScoring) {
	scoreGenesParallel(genePool);
} else {
	scoreGenes(genePool);
}
```

On the topic of Java 8: Parallelization would still be possible to do it without it, but not as clean and easy as with Java 8 features. Also since Java 9 is just around the corner, I would not shy away from 8 any longer ;)
What exactly would make it harder to use that Java 7 on Linux? (I'm asking because I'm developing on Linux at work, and so far had no problems with it)

And no, it didn't sound too negative, as you gave proper explanations to your argument. :)
There are some people who woulds just go `"Bleh, that's stupid. *CLOSED*"`, without further explanation.

NPE

**Comment by nickd3000** [comment]
Hi NPE, 
It seems like most Linux installs don't ship with the repository for Java 8 / openjdk 8 available by default, so anyone wanting to compile the code on linux would need to figure out how to add the repository and then install Java 8. I just confirmed this with a fresh install of Debian on a virtual machine - I don't know why they still haven't updated their apt system yet, It's pretty annoying.

I suppose I could just add something to the readme file saying this requires java 8 and maybe add a link to how to get it, so sure go ahead.

Thats fine to add scoreGenesParallel() to the main Javolver class - I was just thinking to the future where we might want to add some extra parallel functions and keep them separate but what you described should be cool.

Looking forward to the scoring methods using up all of my processors :)

Nick

**Comment by NPException** [comment]
Alrighty, I'll close this request then and prepare a new one now :)

**Pull request closed without merging**